File: leetcode/general/1-two-sum.py
"""
Given an array of integers, return indices of the
two numbers such that they add up to a specific target.

You may assume that each input would have exactly
one solution, and you may not use the same element twice.

Example:

Given nums = [2, 7, 11, 15], target = 9,

Because nums[0] + nums[1] = 2 + 7 = 9,

return [0, 1].
--------------------------------------
- Had notes but deleted old file

- Created sorted version of original
- use high/low pointer in sorted array, decrement or increment til solution is found
- Return array of indices in original array, be sure to catch case like [3,3]
"""


class Solution(object):
    def twoSum(self, nums, target):
        """
        O(n) solution: For each
        number in the array, map
        its complement to its index.
        If we find a value that is the
        complement of some other value,
        return the indices of both.
        (this beat 99.74% of existing solutions)
        """
        map = {}
        for i in range(len(nums)):
            if nums[i] not in map:
                map[target - nums[i]] = i
            else:
                return [map[nums[i]], i]
        return [-1, -1]

    def twoSumLinearithmic(self, nums, target):
        """
        O(nlogn) - sort the array, then maintain
        high and low pointers in it. If arr[hi] + arr[low]
        is greater than the target, decrement hi. If lower,
        increment low. Once the two items are found,
        find their original indices and return them.
        """
        # Create sorted arr
        sortedArr = sorted(nums)
        hi = len(sortedArr) - 1
        lo = 0

        # Get indicies in sorted arr
        while sortedArr[lo] + sortedArr[hi] != target:
            if sortedArr[lo] + sortedArr[hi] > target:
                hi -= 1
            else:
                lo += 1

        # Convert to indices in unsorted arr
        solution = []
        for i in range(len(nums)):
            if nums[i] == sortedArr[lo]:
                solution.append(i)
                break

        for i in range(len(nums)-1, -1, -1):
            if nums[i] == sortedArr[hi]:
                solution.append(i)
                break

        return solution


if __name__ == '__main__':
    s = Solution()
    assert s.twoSum([3, 3], 6) == [0, 1]
    assert s.twoSum([2, 7, 11, 15], 9) == [0, 1]

File: leetcode/1-two-sum/1-two-sum.cpp
/*
Given an array of integers, return indices of the two numbers such that they add
up to a specific target.

You may assume that each input would have exactly one solution, and you may not
use the same element twice.

Example:

Given nums = [2, 7, 11, 15], target = 9,

Because nums[0] + nums[1] = 2 + 7 = 9,
return [0, 1].
----------------------------------------------
In: array of ints and target int
Out: indices of ints that add to that value
- Solution always exists
- Can't use same elements twice
- Not necessarily sorted

- Can we sort elements, and then use the following procedure:
  - start with hi and lo
  - if arr[hi] + arr[lo] > target, decrease hi
  - if arr[hi] + arr[lo] < target, increase lo
  - theoretically (though not for this problem), if hi == lo, no solution.

- Note that we are returning the indices, so we have to somehow preserve the
original ordering.
- We could do faster lookups by sticking things in a map, but that's more mem
intensive
- Or at the end we can look up the element, but would be problematic on a case
like [0,4,4,9] and 8.
  - Maybe we can just do searches but skip if index is the same?
---------------------------------------------------------------------------

// this finds the actual elements
- sort elements
- create answer vector
- hi = len-1
- lo = 0
- while hi != lo (catches empty):
  - if arr[lo] + arr[hi] = target
    - push arr[hi]
    - push arr[lo]
  - elif arr[lo] + arr[hi] > target
    - hi--
  - else:
    - lo--

// then, passed back to the main method:
  - new solution vector
  - first = answer[0]
  - find answer[0] in original vector, return index
  - find answer[1] in original vector not equal to answer[0]
*/

#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

class Solution {
 public:
  vector<int> getValues(vector<int>& nums, int target) {
    vector<int> solution = new vector();
    vector<int> sorted = vector(nums);
    sort(sorted.front, sorted.back);
    while (sorted.front != sorted.back) {
    }
  }

  vector<int> twoSum(vector<int>& nums, int target) {
    vector<int> solution = getValues(nums, target);

    // Find index of second value, replace in values vector
    for (int i = 0; i < nums.size(); i++) {
      if (nums[i] == solution[0]) {
        solution[0] = nums[i];
      }
    }

    // Find index of second value; replace in values vector
    for (int i = 0; i < nums.size(); i++) {
      if (nums[i] == solution[1] && nums[i] != solution[0]) {
        solution[1] = nums[i];
      }
    }

    return solution;
  }
};

int main() {
  cout << "Hello, leetcode!" << endl;
  return 0;
}
